runtime/vm/UpcallVMHelpers.cpp

```diff
@@ -244,20 +242,21 @@ native2InterpJavaUpcallImpl(J9UpcallMetaData *data, void *argsListPointer)
 	returnsObject = (J9NtcPointer == returnType) || (J9NtcStruct == returnType);
 
 	if (buildCallInStackFrameHelper(currentThread, &newELS, returnsObject)) {
+		j9object_t mhMetaData = NULL;
+		j9object_t nativeArgArray = NULL;
 
-		/* Save the current executing method and restore later after the memory allocation
-		 * for pointer/struct in the argument list.
-		 */
-		thrLiterals = currentThread->literals;
+		/* Store the allocated memory objects for the struct/pointer arguments to the java array */
+		if (!storeMemArgObjectsToJavaArray(data, currentThread)) {
+			throwOOM = true;
+			goto done;
+		}
 
+		mhMetaData = J9_JNI_UNWRAP_REFERENCE(data->mhMetaData);
+		nativeArgArray = J9VMOPENJ9INTERNALFOREIGNABIUPCALLMHMETADATA_NATIVEARGARRAY(currentThread, mhMetaData);
 		/* The argument list of the upcall method handle on the stack includes the target method handle,
 		 * the method arguments and the appendix which is set via MethodHandleResolver.upcallLinkCallerMethod().
-		 *
-		 * Note: push the target method handle on the special frame so as to avoid updating the address
-		 * on the stack by GC when allocating memory for the next pointer/struct of the argument list.
 		 */
-		j9object_t calleeHandle = J9VMOPENJ9INTERNALFOREIGNABIUPCALLMHMETADATA_CALLEEMH(currentThread, J9_JNI_UNWRAP_REFERENCE(data->mhMetaData));
-		PUSH_OBJECT_IN_SPECIAL_FRAME(currentThread, calleeHandle);
+		*(j9object_t*)--currentThread->sp = J9VMOPENJ9INTERNALFOREIGNABIUPCALLMHMETADATA_CALLEEMH(currentThread, mhMetaData);
```

Comment:
Though it's the same in practice, I would prefer that the cast (to UDATA) be on the source rather than the target

Reply:
Ignore - the code as it is lines up with the I_32 cases.